woltka/workflow.py

```diff
@@ -0,0 +1,590 @@
+        with readzip(fp) as f:
+            map_ = read_map(f)
+        update_dict(tree, map_)
+        if map_rank:
+            update_dict(rankd, {k: rank for k in set(map_.values())})
+
+    # fill root
+    root = fill_root(tree)
+
+    if is_build:
+        click.echo(' Done.')
+        click.echo(f'Total classification units: {len(tree)}.')
+    return tree, rankd, named, root
+
+
+def reshape_rmap(rmap:    dict,
+                 deidx:   bool = None,
+                 demux:   bool = None,
+                 samples: list = None,
+                 files:   dict = None,
+                 fp:      str = None) -> dict:
+    """Reshape a read map.
+
+    Parameters
+    ----------
+    rmap : dict
+        Read map to manipulate.
+    deidx : bool, optional
+        Strip suffixes from subject IDs.
+    demux : bool, optional
+        Whether perform demultiplexing.
+    samples : list of str, optional
+        Sample ID list to include.
+    files : list or dict
+        Map of filepaths to sample IDs.
+    fp : str
+        Path to current alignment file.
+
+    Returns
+    -------
+    dict
+        Reshaped read map.
+    """
+    # strip indices from subjects
+    if deidx:
+        strip_index(rmap)
+
+    # demultiplex into multiple samples
+    if demux:
+        return demultiplex(rmap, samples)
+
+    # sample Id from filename
+    else:
+        return {files[fp]: rmap}
+
+
+def assign_rmap(rmap:     dict,
+                data:     dict,
+                rank:     str,
+                sample:   str,
+                rank2dir: dict = None,
+                tree:     dict = None,
+                rankd:    dict = None,
+                named:    dict = None,
+                root:      str = None,
+                above:    bool = False,
+                major:     int = None,
+                ambig:     str = True,
+                subok:    bool = None):
+    """Assign query sequences in a query-to-subjects map to classification
+    units based on their subjects.
+
+    Parameters
+    ----------
+    rmap : dict
+        Read map to assign.
+    data : dict
+        Master data structure.
+    rank : str
+        Target rank to assign to.
+    sample : str
+        Sample ID.
+    rank2dir : dict, optional
+        Directory of output maps per rank.
+
+    tree : dict, optional
+        Hierarchical classification system.
+    rankd : dict, optional
+        Rank dictionary.
+    named : dict
+        Taxon name directory.
+    root : str, optional
+        Root identifier.
+
+    above : bool, optional
+        Assignment above given rank is acceptable (for fixed ranks).
+    major : float, optional
+        Majority-rule assignment threshold (available only with a fixed rank
+        and not above or ambig).
+    ambig : bool, optional
+        Count occurrence of each possible assignment instead of targeting one
+        assignment (available only with a fixed rank and not above).
+    subok : bool, optional
```

Comment:
This name is unhelpful.

Reply:
I agree that "subok" is not a good parameter. I attempted find a phrase to describe "must move up classification hierarchy by at least one step; cannot directly report subject" and "subok" was unfortunately the best I could get...

Also, I am thinking of making this parameter "smart": in default status (`None`), the program will determine based on the input classification files...

Reply:
ahhh, see I was not able to register "sub-oh-kay" from `subok`, (I thought "sub-awk"). Maybe something like "subject_ok". Though I think this has a different meaning "cannot directly report subject" unless that is the behavior when this variable is False? Could not find documentation in the README, so trying to triangulate from cli.py

Reply:
I would also like to reiterate that I __recommend against__ so-called "smart parameters". I think it is more reasonable to expect a user to know their needs, than it is to expect a user to be intimately familiar with the way that your tool makes "auto" decisions. Lines in the readme that explain 
> this flag does <this>
will be easier for a user to use/navigate than 
> this flag does <this> if you have one type of input and <that> if you have another type of input unless it is <None> then it will be inferred based on the contents of the file using some rules
Especially when they have to navigate 5 of these sorts of flags :)